### ISSUE TITLE
REP Socket blocking

### ISSUE THREAD

**Issue by ylegoc** [zeromq/jeromq#411]
I'm facing an issue that is maybe trivial to solve (my knowledge of JeroMQ may be limited).
I have a server that receives requests through a REP socket.
<pre>
Socket server = context.createSocket(ZMQ.REP);
server.bind("...");

while (true) {
	ZMsg message = null;
	try {
		message = ZMsg.recvMsg(server);
		...
</pre>
Sometimes the recvMsg call blocks whereas a request has been sent. A way to unblock it is to send another request. Then the recvMsg returns and the first request can be processed (as well as the second one). I have not this problem with JZMQ. 
Note that the number of requests correctly processed before the blocking is small (10 - 100).

**Comment by ylegoc** [comment]
In fact I did not say that there were other PUB messages that were sent. I tried to debug by using the code of the 0.3.7-SNAPSHOT version and here is the log. "Receiving request" is written before the reception of the request. "Processing request" is written after. It appears that: 
- In case of a normal execution, the signaler associated to the socket sends PIPE_TERM, ACTIVATE_READ, PIPE_TERM_ACK and ACTIVATE_READ.
- In case of blocking, the signaler sends PIPE_TERM, ACTIVATE_READ, PIPE_TERM_ACK but not the last ACTIVATE_READ.
It seems that it's because the Poller is blocked and doesn't trigger the ACTIVATE_READ.
The log shows two requests, the first succeeds, the second blocks.
<pre>
Receiving request
	zmq.Signaler@728938a9 before select
	Poller zmq.Poller@23b956f8 selects ok
	zmq.Signaler@728938a9 after select
	zmq.Signaler@728938a9 before select
		zmq.Signaler@31221be2 send Cmd[zmq.Pub$PubSession@701e499a[2], 2, PLUG] false
	zmq.Signaler@31221be2 Signaler sends
		zmq.Signaler@31221be2 send Cmd[zmq.TcpListener@5910e440, 2, OWN, zmq.Pub$PubSession@701e499a[2]] true
		zmq.Signaler@31221be2 send Cmd[zmq.Pub$PubSession@701e499a[2], 2, ATTACH, zmq.StreamEngine@3a671e8c] true
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@31221be2 send Cmd[zmq.Rep$RepSession@139cb2e0[1], 2, PLUG] true
		zmq.Signaler@31221be2 send Cmd[zmq.TcpListener@377dca04, 2, OWN, zmq.Rep$RepSession@139cb2e0[1]] true
		zmq.Signaler@31221be2 send Cmd[zmq.Rep$RepSession@139cb2e0[1], 2, ATTACH, zmq.StreamEngine@7e6fded2] true
		zmq.Signaler@6267c3bb send Cmd[zmq.Pub@533ddba[2], 4, BIND, zmq.Pipe@7c3f1a1c[zmq.Pub@533ddba[2]]] true
		zmq.Signaler@728938a9 send Cmd[zmq.Rep@21b8d17c[1], 3, BIND, zmq.Pipe@8d35efe[zmq.Rep@21b8d17c[1]]] false
	zmq.Signaler@728938a9 Signaler sends
	zmq.Signaler@728938a9 after select
		zmq.Signaler@31221be2 send Cmd[zmq.TcpListener@377dca04, 2, TERM_REQ, zmq.Rep$RepSession@6413ddde[1]] false
	zmq.Signaler@31221be2 Signaler sends
	zmq.Signaler@728938a9 before select
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@31221be2 send Cmd[zmq.Rep$RepSession@6413ddde[1], 2, TERM, -1] true
		zmq.Signaler@728938a9 send Cmd[zmq.Pipe@4a20026f[zmq.Rep@21b8d17c[1]], 3, PIPE_TERM] false
	zmq.Signaler@728938a9 Signaler sends
	zmq.Signaler@728938a9 after select
		zmq.Signaler@728938a9 send Cmd[zmq.Pipe@4a20026f[zmq.Rep@21b8d17c[1]], 3, ACTIVATE_READ] true
		zmq.Signaler@31221be2 send Cmd[zmq.Pipe@3a4afd8d[zmq.Rep$RepSession@6413ddde[1]], 2, PIPE_TERM_ACK] false
	Poller zmq.Poller@23b956f8 selects
	zmq.Signaler@31221be2 Signaler sends
	zmq.Signaler@728938a9 before select
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@31221be2 send Cmd[zmq.TcpListener@377dca04, 2, TERM_ACK] true
		zmq.Signaler@728938a9 send Cmd[zmq.Pipe@4a20026f[zmq.Rep@21b8d17c[1]], 3, PIPE_TERM_ACK] false
	zmq.Signaler@728938a9 Signaler sends
	zmq.Signaler@728938a9 after select
	Poller zmq.Poller@23b956f8 selects
	zmq.Signaler@728938a9 before select
	Poller zmq.Poller@23b956f8 selects ok
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@728938a9 send Cmd[zmq.Pipe@8d35efe[zmq.Rep@21b8d17c[1]], 3, ACTIVATE_READ] false
	zmq.Signaler@728938a9 Signaler sends
	zmq.Signaler@728938a9 after select
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
Processing request
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
	Poller zmq.Poller@23b956f8 selects
		zmq.Signaler@31221be2 send Cmd[zmq.Pipe@2db0f6b2[zmq.Pub$PubSession@38d536e2[2]], 2, ACTIVATE_READ] false
	zmq.Signaler@31221be2 Signaler sends
		zmq.Signaler@31221be2 send Cmd[zmq.Pipe@2e5c649[zmq.Pub$PubSession@2b54d336[2]], 2, ACTIVATE_READ] true
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@31221be2 send Cmd[zmq.Pipe@2aae9190[zmq.Pub$PubSession@73a74bb6[2]], 2, ACTIVATE_READ] true
		zmq.Signaler@31221be2 send Cmd[zmq.Pipe@1996cd68[zmq.Rep$RepSession@139cb2e0[1]], 2, ACTIVATE_READ] true
		zmq.Signaler@728938a9 send Cmd[zmq.Pipe@8d35efe[zmq.Rep@21b8d17c[1]], 3, ACTIVATE_WRITE, 0] false
	zmq.Signaler@728938a9 Signaler sends
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@6267c3bb send Cmd[zmq.Pipe@7c3f1a1c[zmq.Pub@533ddba[2]], 4, ACTIVATE_READ] false
	zmq.Signaler@6267c3bb Signaler sends
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@31221be2 send Cmd[zmq.TcpListener@377dca04, 2, TERM_REQ, zmq.Rep$RepSession@139cb2e0[1]] false
	zmq.Signaler@31221be2 Signaler sends
	Poller zmq.Poller@23b956f8 selects
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@31221be2 send Cmd[zmq.Rep$RepSession@139cb2e0[1], 2, TERM, -1] true
		zmq.Signaler@728938a9 send Cmd[zmq.Pipe@8d35efe[zmq.Rep@21b8d17c[1]], 3, PIPE_TERM] true
	Poller zmq.Poller@23b956f8 selects
Receiving request
	zmq.Signaler@728938a9 before select
	zmq.Signaler@728938a9 after select
		zmq.Signaler@31221be2 send Cmd[zmq.Pipe@1996cd68[zmq.Rep$RepSession@139cb2e0[1]], 2, PIPE_TERM_ACK] false
	zmq.Signaler@31221be2 Signaler sends
	zmq.Signaler@728938a9 before select
	Poller zmq.Poller@23b956f8 selects ok
		zmq.Signaler@31221be2 send Cmd[zmq.TcpListener@377dca04, 2, TERM_ACK] true
		zmq.Signaler@728938a9 send Cmd[zmq.Pipe@8d35efe[zmq.Rep@21b8d17c[1]], 3, PIPE_TERM_ACK] false
	zmq.Signaler@728938a9 Signaler sends
	Poller zmq.Poller@23b956f8 selects
	zmq.Signaler@728938a9 after select
	zmq.Signaler@728938a9 before select
</pre>

I can test deeper if somebody tells me where to look at.

**Comment by daveyarwood** [comment]
@ylegoc Apologies for the extremely late response.

In case it wasn't obvious: `ZMsg.recvMsg` (like `ZMQ.Socket.recv`) will block until a message is available by default. If you need to do things concurrently (e.g. if you need to handle multiple sockets in a loop), the thing to do is to use a ZLoop or a poller.

Beyond that, I don't think I understand enough about the issue you were seeing to be able to help much. Maybe it would help if you re-summarized the issue?

**Comment by ylegoc** [comment]
Hi Dave,

No problem for your late answer :).

So to be simple, I have a server application that is using a REP socket to process requests.
Sometimes, after a client sent a request A using its REQ socket connected to the REP socket, the server REP socket remains blocked indefinitely in the recvMsg method.

At this point, if a new request B is sent to the REP socket, the recvMsg returns twice: one for the request A and one for the request B.

Note that the number and frequency of requests sent by the client is reasonable.
As the use of the REQ/REP socket is trivial and has been tested in multiple cases, I suppose that the problem comes from the context. There are other PUB sockets publishing messages in the server application.
And when I looked deeper into the code, I noticed a correlation between the trigger order and the request that is "blocked" (cf my second message).

I hope it is clearer.

**Comment by daveyarwood** [comment]
Much clearer, thanks!

Are you able to to rule out whether or not the PUB sockets are causing the issue? i.e. if you are able to not create/use the PUB sockets at all and just do the REQ/REP part, is the issue still there?

**Comment by ylegoc** [comment]
Hi Dave,

No problem for your late answer :).

So to be simple, I have a server application that is using a REP socket to process requests.
Sometimes, after a client sent a request A using its REQ socket connected to the REP socket, the server REP socket remains blocked indefinitely in the recvMsg method.

At this point, if a new request B is sent to the REP socket, the recvMsg returns twice: one for the request A and one for the request B.

Note that the number and frequency of requests sent by the client is reasonable.
As the use of the REQ/REP socket is trivial and has been tested in multiple cases, I suppose that the problem comes from the context. There are other PUB sockets publishing messages in the server application.
And when I looked deeper into the code, I noticed a correlation between the trigger order and the request that is "blocked" (cf my second message).

I hope it is clearer.
